File: lib/cosmetics.ts
/**
 * Cosmetics: profile cards, titles, and badge items (FACEIT-style inventory).
 *
 * The Discord bot owns the catalog and grant/revoke (core/cosmetics.py — the
 * schema below is mirrored there in core/schema.py); the website owns the
 * equip/unequip flow and rendering. Items are keyed to players by player name.
 *
 * Equipped state is a flag on the inventory row, so a bot-side revoke (DELETE)
 * atomically removes the equip too. Single-equip for card/title is enforced at
 * write time; badges allow up to MAX_EQUIPPED_BADGES equipped at once.
 *
 * The equipped title is mirrored into players.title so the bot's /checkplayer
 * embed keeps showing it.
 */

import { client, ensurePlayerCoinsColumn } from "@/lib/db";

export type CosmeticType = "card" | "title" | "badge" | "frame";

export interface InventoryItem {
  id: number; // catalog item id (what equip endpoints take)
  slug: string;
  type: CosmeticType;
  name: string;
  description: string;
  asset: string | null;
  category: string | null;
  season: string | null;
  rarity: string;
  grantedAt: number;
  equipped: boolean;
}

/** Equipped cosmetics as rendered on a public profile. */
export interface ProfileCosmetics {
  card: { slug: string; name: string; asset: string | null } | null;
  frame: { slug: string; name: string; asset: string | null } | null;
  title: string | null;
  badges: { slug: string; name: string; description: string; asset: string | null }[];
}

export const MAX_EQUIPPED_BADGES = 5;

let schemaReady: Promise<void> | null = null;

/** Create the cosmetics tables once per process (idempotent, mirrors
 *  core/schema.py). Schema only — item seeding is bot-side. */
function ensureCosmeticsSchema(): Promise<void> {
  if (!schemaReady) {
    schemaReady = (async () => {
      await client.batch([
        `CREATE TABLE IF NOT EXISTS cosmetic_items (
           id          INTEGER PRIMARY KEY AUTOINCREMENT,
           slug        TEXT NOT NULL UNIQUE,
           type        TEXT NOT NULL,
           name        TEXT NOT NULL,
           description TEXT DEFAULT '',
           asset       TEXT DEFAULT NULL,
           category    TEXT DEFAULT NULL,
           season      TEXT DEFAULT NULL,
           rarity      TEXT DEFAULT 'common',
           created_at  INTEGER,
           price       INTEGER DEFAULT 0 )`,
        `CREATE TABLE IF NOT EXISTS cosmetic_inventory (
           id          INTEGER PRIMARY KEY AUTOINCREMENT,
           player_name TEXT NOT NULL,
           item_id     INTEGER NOT NULL,
           granted_by  TEXT DEFAULT NULL,
           granted_at  INTEGER,
           equipped    INTEGER NOT NULL DEFAULT 0,
           equipped_at INTEGER )`,
        `CREATE UNIQUE INDEX IF NOT EXISTS idx_cosmetic_inventory_unique
           ON cosmetic_inventory(player_name, item_id)`,
        `CREATE INDEX IF NOT EXISTS idx_cosmetic_inventory_player
           ON cosmetic_inventory(player_name)`,
      ]);
      // Back-fill the shop price column on DBs created before it existed.
      await client
        .execute("ALTER TABLE cosmetic_items ADD COLUMN price INTEGER DEFAULT 0")
        .catch(() => undefined); // already exists — fine
    })();
  }
  return schemaReady;
}

function rowToItem(r: Record<string, unknown>): InventoryItem {
  return {
    id: Number(r.id),
    slug: r.slug as string,
    type: r.type as CosmeticType,
    name: r.name as string,
    description: (r.description as string) ?? "",
    asset: (r.asset as string) ?? null,
    category: (r.category as string) ?? null,
    season: (r.season as string) ?? null,
    rarity: (r.rarity as string) || "common",
    grantedAt: Number(r.granted_at ?? 0),
    equipped: Number(r.equipped) === 1,
  };
}

/** Everything the player owns, catalog details included. */
export async function getInventory(playerName: string): Promise<InventoryItem[]> {
  await ensureCosmeticsSchema();
  const rs = await client.execute({
    sql: `SELECT i.id, i.slug, i.type, i.name, i.description, i.asset, i.category,
                 i.season, i.rarity, inv.granted_at, inv.equipped
          FROM cosmetic_inventory inv JOIN cosmetic_items i ON i.id = inv.item_id
          WHERE inv.player_name = ?
          ORDER BY i.type, inv.granted_at DESC`,
    args: [playerName],
  });
  return (rs.rows as unknown as Record<string, unknown>[]).map(rowToItem);
}

/**
 * Equip or unequip an owned item. Card/title are single-equip (equipping one
 * unequips the others of that type); the equipped title mirrors into
 * players.title. Badges are capped at MAX_EQUIPPED_BADGES.
 */
export async function setEquipped(
  playerName: string,
  itemId: number,
  equip: boolean
): Promise<"ok" | "not_owned" | "badge_limit"> {
  await ensureCosmeticsSchema();
  const rs = await client.execute({
    sql: `SELECT i.type, i.name FROM cosmetic_inventory inv
          JOIN cosmetic_items i ON i.id = inv.item_id
          WHERE inv.player_name = ? AND inv.item_id = ?`,
    args: [playerName, itemId],
  });
  if (rs.rows.length === 0) return "not_owned";
  const type = rs.rows[0].type as CosmeticType;
  const name = rs.rows[0].name as string;
  const now = Date.now();

  if (!equip) {
    const stmts = [
      {
        sql: "UPDATE cosmetic_inventory SET equipped = 0, equipped_at = NULL WHERE player_name = ? AND item_id = ?",
        args: [playerName, itemId] as (string | number)[],
      },
    ];
    if (type === "title") {
      stmts.push({ sql: "UPDATE players SET title = '' WHERE name = ?", args: [playerName] });
    }
    await client.batch(stmts);
    return "ok";
  }

  if (type === "badge") {
    const countRs = await client.execute({
      sql: `SELECT COUNT(*) AS c FROM cosmetic_inventory inv
            JOIN cosmetic_items i ON i.id = inv.item_id
            WHERE inv.player_name = ? AND i.type = 'badge' AND inv.equipped = 1
              AND inv.item_id != ?`,
      args: [playerName, itemId],
    });
    if (Number(countRs.rows[0]?.c ?? 0) >= MAX_EQUIPPED_BADGES) return "badge_limit";
    await client.execute({
      sql: "UPDATE cosmetic_inventory SET equipped = 1, equipped_at = ? WHERE player_name = ? AND item_id = ?",
      args: [now, playerName, itemId],
    });
    return "ok";
  }

  // card / title: unequip every other item of the type, then equip the target.
  const stmts = [
    {
      sql: `UPDATE cosmetic_inventory SET equipped = 0, equipped_at = NULL
            WHERE player_name = ? AND item_id IN
              (SELECT id FROM cosmetic_items WHERE type = ?)`,
      args: [playerName, type] as (string | number)[],
    },
    {
      sql: "UPDATE cosmetic_inventory SET equipped = 1, equipped_at = ? WHERE player_name = ? AND item_id = ?",
      args: [now, playerName, itemId],
    },
  ];
  if (type === "title") {
    // Mirror into players.title so the bot's /checkplayer shows it.
    stmts.push({ sql: "UPDATE players SET title = ? WHERE name = ?", args: [name, playerName] });
  }
  await client.batch(stmts);
  return "ok";
}

/** Equipped card + frame assets for one player, as used by list/lobby views. */
export interface EquippedVisuals {
  card: string | null;
  frame: string | null;
}

/** player_name → equipped card/frame assets, for list views (single query). */
export async function getEquippedVisualsMap(): Promise<Map<string, EquippedVisuals>> {
  await ensureCosmeticsSchema();
  const rs = await client.execute(
    `SELECT inv.player_name AS name, i.type AS type, i.asset AS asset
     FROM cosmetic_inventory inv JOIN cosmetic_items i ON i.id = inv.item_id
     WHERE inv.equipped = 1 AND i.type IN ('card', 'frame') AND i.asset IS NOT NULL`
  );
  const map = new Map<string, EquippedVisuals>();
  for (const r of rs.rows as unknown as { name: string; type: string; asset: string }[]) {
    const v = map.get(r.name) ?? { card: null, frame: null };
    if (r.type === "card" && !v.card) v.card = r.asset;
    if (r.type === "frame" && !v.frame) v.frame = r.asset;
    map.set(r.name, v);
  }
  return map;
}

/** The equipped cosmetics for a public profile (card, title text, ≤5 badges). */
export async function getEquippedCosmetics(playerName: string): Promise<ProfileCosmetics> {
  await ensureCosmeticsSchema();
  const rs = await client.execute({
    sql: `SELECT i.slug, i.type, i.name, i.description, i.asset
          FROM cosmetic_inventory inv JOIN cosmetic_items i ON i.id = inv.item_id
          WHERE inv.player_name = ? AND inv.equipped = 1
          ORDER BY inv.equipped_at ASC`,
    args: [playerName],
  });
  const out: ProfileCosmetics = { card: null, frame: null, title: null, badges: [] };
  for (const r of rs.rows as unknown as Record<string, unknown>[]) {
    const type = r.type as CosmeticType;
    if (type === "card" && !out.card) {
      out.card = {
        slug: r.slug as string,
        name: r.name as string,
        asset: (r.asset as string) ?? null,
      };
    } else if (type === "frame" && !out.frame) {
      out.frame = {
        slug: r.slug as string,
        name: r.name as string,
        asset: (r.asset as string) ?? null,
      };
    } else if (type === "title" && !out.title) {
      out.title = r.name as string;
    } else if (type === "badge" && out.badges.length < MAX_EQUIPPED_BADGES) {
      out.badges.push({
        slug: r.slug as string,
        name: r.name as string,
        description: (r.description as string) ?? "",
        asset: (r.asset as string) ?? null,
      });
    }
  }
  return out;
}

// --- shop ------------------------------------------------------------------

/** A purchasable catalog item, plus whether this player already owns it. */
export interface ShopItem {
  id: number;
  slug: string;
  type: CosmeticType;
  name: string;
  description: string;
  asset: string | null;
  rarity: string;
  price: number;
  owned: boolean;
}

/** Everything for the shop page: the player's balance + the purchasable
 *  catalog (price > 0), each flagged with whether they already own it. */
export async function getShop(
  playerName: string
): Promise<{ coins: number; items: ShopItem[] }> {
  await ensureCosmeticsSchema();
  await ensurePlayerCoinsColumn();
  const [balanceRs, itemsRs] = await Promise.all([
    client.execute({ sql: "SELECT coins FROM players WHERE name = ?", args: [playerName] }),
    client.execute({
      sql: `SELECT i.id, i.slug, i.type, i.name, i.description, i.asset, i.rarity, i.price,
                   EXISTS(SELECT 1 FROM cosmetic_inventory inv
                          WHERE inv.player_name = ? AND inv.item_id = i.id) AS owned
            FROM cosmetic_items i
            WHERE i.price > 0
            ORDER BY i.price ASC, i.type`,
      args: [playerName],
    }),
  ]);
  const coins = Number(balanceRs.rows[0]?.coins ?? 0);
  const items = (itemsRs.rows as unknown as Record<string, unknown>[]).map((r) => ({
    id: Number(r.id),
    slug: r.slug as string,
    type: r.type as CosmeticType,
    name: r.name as string,
    description: (r.description as string) ?? "",
    asset: (r.asset as string) ?? null,
    rarity: (r.rarity as string) || "common",
    price: Number(r.price ?? 0),
    owned: Number(r.owned) === 1,
  }));
  return { coins, items };
}

export type PurchaseResult =
  | { status: "ok"; coins: number }
  | { status: "not_for_sale" | "already_owned" | "insufficient" | "no_such_item" | "no_such_player"; coins: number };

/**
 * Buy a catalog item with HL Coins. Atomicity without a real transaction:
 * deduct with a `WHERE coins >= price` guard (so a concurrent/insufficient
 * spend affects 0 rows), then insert the inventory row; if that insert loses a
 * race (unique index → 0 rows), refund the coins. Balances can't go negative.
 */
export async function purchaseItem(
  playerName: string,
  itemId: number
): Promise<PurchaseResult> {
  await ensureCosmeticsSchema();
  await ensurePlayerCoinsColumn();

  const itemRs = await client.execute({
    sql: "SELECT price FROM cosmetic_items WHERE id = ?",
    args: [itemId],
  });
  if (itemRs.rows.length === 0) return { status: "no_such_item", coins: 0 };
  const price = Number(itemRs.rows[0].price ?? 0);

  const balRs = await client.execute({
    sql: "SELECT coins FROM players WHERE name = ?",
    args: [playerName],
  });
  if (balRs.rows.length === 0) return { status: "no_such_player", coins: 0 };
  const coins = Number(balRs.rows[0].coins ?? 0);

  if (price <= 0) return { status: "not_for_sale", coins };

  const alreadyRs = await client.execute({
    sql: "SELECT 1 FROM cosmetic_inventory WHERE player_name = ? AND item_id = ?",
    args: [playerName, itemId],
  });
  if (alreadyRs.rows.length > 0) return { status: "already_owned", coins };
  if (coins < price) return { status: "insufficient", coins };

  // Guarded deduct: 0 rows means the balance changed under us / went too low.
  const deduct = await client.execute({
    sql: "UPDATE players SET coins = coins - ? WHERE name = ? AND coins >= ?",
    args: [price, playerName, price],
  });
  if (deduct.rowsAffected === 0) {
    const fresh = Number((await client.execute({ sql: "SELECT coins FROM players WHERE name = ?", args: [playerName] })).rows[0]?.coins ?? 0);
    return { status: "insufficient", coins: fresh };
  }

  const insert = await client.execute({
    sql: `INSERT OR IGNORE INTO cosmetic_inventory (player_name, item_id, granted_by, granted_at)
          VALUES (?, ?, 'shop', ?)`,
    args: [playerName, itemId, Date.now()],
  });
  if (insert.rowsAffected === 0) {
    // Lost the ownership race — refund the deducted coins.
    await client.execute({
      sql: "UPDATE players SET coins = coins + ? WHERE name = ?",
      args: [price, playerName],
    });
    return { status: "already_owned", coins };
  }

  return { status: "ok", coins: coins - price };
}

File: core/cosmetics.py
"""Cosmetic items: profile cards, titles, and badges (the FACEIT-style
inventory shared with the website).

Pure DB logic — no Discord imports — so it's callable from admin commands,
scheduled tasks, the season reset, and (later) event code. The website mirrors
the schema and owns the equip/unequip flow (HL website .../lib/cosmetics.ts);
this module owns the catalog and grant/revoke paths.

Tables (created in core.schema.ensure_schema):
    cosmetic_items      the catalog (slug, type card|title|badge, name, asset…)
    cosmetic_inventory  who owns what + equipped flag

Timestamps are epoch-ms ints (never pass a raw datetime — libsql serializes it
to int-ms anyway, so we store ints deliberately and consistently).
"""

import logging
import re
import time

from config.settings import TOP10_COUNT, TOP10_MIN_ELO
from core import db

logger = logging.getLogger(__name__)

# Slug for the dynamic "currently in the top 10" badge (seeded at startup).
TOP10_BADGE_SLUG = "top10-current"

# Season placement tiers, best first: (max standing, tier label). A player gets
# only the best tier they reach — #1 gets Champion (not also Top 3 / Top 10).
PLACEMENT_TIERS = [(1, "Champion"), (3, "Top 3"), (10, "Top 10")]

_SLUG_RE = re.compile(r"^[a-z0-9-]{2,40}$")


def now_ms() -> int:
    return int(time.time() * 1000)


def valid_slug(slug: str) -> bool:
    return bool(_SLUG_RE.match(slug))


def slugify(text: str) -> str:
    """Free text -> catalog slug ('Season 1' -> 'season-1')."""
    return re.sub(r"-+", "-", re.sub(r"[^a-z0-9-]", "-", text.strip().lower())).strip("-")


async def ensure_item(
    slug: str,
    item_type: str,
    name: str,
    description: str = "",
    asset: str | None = None,
    category: str | None = None,
    season: str | None = None,
    rarity: str = "common",
    price: int = 0,
) -> int | None:
    """Create a catalog item if it doesn't exist (idempotent by slug).
    ``price`` > 0 lists the item in the website shop. Returns the item id, or
    None if the insert raced and the lookup failed."""
    await db.execute(
        """INSERT OR IGNORE INTO cosmetic_items
           (slug, type, name, description, asset, category, season, rarity, created_at, price)
           VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)""",
        (slug, item_type, name, description, asset, category, season, rarity, now_ms(), max(0, price)),
    )
    row = await db.fetchone("SELECT id FROM cosmetic_items WHERE slug = ?", (slug,))
    return row[0] if row else None


async def set_price(slug: str, price: int) -> bool:
    """Set an item's shop price (0 = not for sale). Returns False if no such item."""
    if await get_item(slug) is None:
        return False
    await db.execute(
        "UPDATE cosmetic_items SET price = ? WHERE slug = ?", (max(0, price), slug)
    )
    return True


async def get_coins(player_name: str) -> int | None:
    """A player's HL Coin balance, or None if the player doesn't exist."""
    row = await db.fetchone("SELECT coins FROM players WHERE name = ?", (player_name,))
    return int(row[0] or 0) if row else None


async def grant_coins(player_name: str, amount: int) -> int | None:
    """Add ``amount`` HL Coins to a player (negative to remove; balance floored
    at 0). Returns the new balance, or None if the player doesn't exist. This is
    the ONLY source of coins for now."""
    current = await get_coins(player_name)
    if current is None:
        return None
    new_balance = max(0, current + amount)
    await db.execute(
        "UPDATE players SET coins = ? WHERE name = ?", (new_balance, player_name)
    )
    return new_balance


async def ensure_builtin_items() -> None:
    """Seed catalog items the system grants automatically. Idempotent; called
    from ensure_schema() on every startup."""
    await ensure_item(
        TOP10_BADGE_SLUG,
        "badge",
        "Top 10",
        description="Currently in the season's top 10",
        asset=f"/badgeicons/{TOP10_BADGE_SLUG}.png",
        category="dynamic",
        rarity="legendary",
    )


async def get_item(slug: str):
    """Catalog row (id, slug, type, name, description, asset, category, season,
    rarity) or None."""
    return await db.fetchone(
        "SELECT id, slug, type, name, description, asset, category, season, rarity "
        "FROM cosmetic_items WHERE slug = ?",
        (slug,),
    )


async def list_items(item_type: str | None = None):
    """All catalog rows, optionally filtered by type, newest first."""
    if item_type:
        return await db.fetchall(
            "SELECT slug, type, name, description, category, season, rarity "
            "FROM cosmetic_items WHERE type = ? ORDER BY id DESC",
            (item_type,),
        )
    return await db.fetchall(
        "SELECT slug, type, name, description, category, season, rarity "
        "FROM cosmetic_items ORDER BY id DESC"
    )


async def get_inventory(player_name: str):
    """A player's items: (slug, type, name, rarity, equipped, granted_by)."""
    return await db.fetchall(
        """SELECT i.slug, i.type, i.name, i.rarity, inv.equipped, inv.granted_by
           FROM cosmetic_inventory inv JOIN cosmetic_items i ON i.id = inv.item_id
           WHERE inv.player_name = ? ORDER BY i.type, inv.granted_at DESC""",
        (player_name,),
    )


async def grant_item(player_name: str, slug: str, granted_by: str) -> str:
    """Grant a catalog item to a player. This is THE grant path — admin
    commands, top-10 sync, season resets, and future events all call it.

    Returns "granted" | "already_owned" | "no_such_item" | "no_such_player".
    """
    item = await get_item(slug)
    if item is None:
        return "no_such_item"
    player = await db.fetchone("SELECT 1 FROM players WHERE name = ?", (player_name,))
    if player is None:
        return "no_such_player"
    owned = await db.fetchone(
        "SELECT 1 FROM cosmetic_inventory WHERE player_name = ? AND item_id = ?",
        (player_name, item[0]),
    )
    if owned:
        return "already_owned"
    await db.execute(
        """INSERT OR IGNORE INTO cosmetic_inventory
           (player_name, item_id, granted_by, granted_at) VALUES (?, ?, ?, ?)""",
        (player_name, item[0], granted_by, now_ms()),
    )
    return "granted"


async def revoke_item(player_name: str, slug: str) -> bool:
    """Take an item away. Deleting the inventory row atomically removes its
    equipped state too; if it was the player's equipped *title*, also clear the
    players.title mirror so /checkplayer stops showing it."""
    item = await get_item(slug)
    if item is None:
        return False
    row = await db.fetchone(
        "SELECT equipped FROM cosmetic_inventory WHERE player_name = ? AND item_id = ?",
        (player_name, item[0]),
    )
    if row is None:
        return False
    stmts = [(
        "DELETE FROM cosmetic_inventory WHERE player_name = ? AND item_id = ?",
        (player_name, item[0]),
    )]
    if item[2] == "title" and row[0]:
        stmts.append(("UPDATE players SET title = '' WHERE name = ?", (player_name,)))
    await db.batch(stmts)
    return True


async def delete_item(slug: str) -> bool:
    """Remove an item from the catalog and every inventory. Clears the
    players.title mirror for anyone who had it equipped as a title."""
    item = await get_item(slug)
    if item is None:
        return False
    stmts = []
    if item[2] == "title":
        holders = await db.fetchall(
            "SELECT player_name FROM cosmetic_inventory WHERE item_id = ? AND equipped = 1",
            (item[0],),
        )
        for (holder,) in holders:
            stmts.append(("UPDATE players SET title = '' WHERE name = ?", (holder,)))
    stmts.append(("DELETE FROM cosmetic_inventory WHERE item_id = ?", (item[0],)))
    stmts.append(("DELETE FROM cosmetic_items WHERE id = ?", (item[0],)))
    await db.batch(stmts)
    return True


async def sync_top10_badge() -> None:
    """Reconcile the dynamic Top 10 badge with the live standings: grant it to
    players entering the top 10, delete it from players who dropped out.

    Ordering matches cogs.roles.refresh_top10_roles, with one deliberate
    divergence: we additionally require elo > 0, so a fresh season reset (all
    Elo zeroed) doesn't hand the badge to 10 arbitrary players.
    """
    rows = await db.fetchall(
        "SELECT name, elo FROM players ORDER BY elo DESC, matches_won DESC LIMIT ?",
        (TOP10_COUNT,),
    )
    top_set = {r[0] for r in rows if r[1] >= TOP10_MIN_ELO and r[1] > 0}

    item = await get_item(TOP10_BADGE_SLUG)
    if item is None:  # seed missing (shouldn't happen after ensure_schema)
        await ensure_builtin_items()
        item = await get_item(TOP10_BADGE_SLUG)
        if item is None:
            return
    holders = {
        r[0]
        for r in await db.fetchall(
            "SELECT player_name FROM cosmetic_inventory WHERE item_id = ?", (item[0],)
        )
    }

    for name in top_set - holders:
        await grant_item(name, TOP10_BADGE_SLUG, "system:top10")
    dropped = holders - top_set
    if dropped:
        await db.batch([
            (
                "DELETE FROM cosmetic_inventory WHERE player_name = ? AND item_id = ?",
                (name, item[0]),
            )
            for name in dropped
        ])


async def award_season_placements(season_name: str) -> list[str]:
    """At season reset: grant each top player the badge for the best placement
    tier they reached (see PLACEMENT_TIERS). Creates the season's catalog items
    on the fly. Must run while the standings are still live (before the Elo
    reset). Uses /resetdb's exact ordering. Returns "Name — Tier" lines for the
    reset embed.
    """
    max_n = max(n for n, _ in PLACEMENT_TIERS)
    standings = [
        r[0]
        for r in await db.fetchall(
            """SELECT name FROM players WHERE elo > 0
               ORDER BY elo DESC, matches_won DESC, matches_played ASC LIMIT ?""",
            (max_n,),
        )
    ]

    season_slug = slugify(season_name) or "season"
    awarded = []
    for pos, player_name in enumerate(standings, start=1):
        # Best (first) tier whose cutoff covers this standing.
        tier_label = next(label for n, label in PLACEMENT_TIERS if pos <= n)
        slug = f"{season_slug}-{slugify(tier_label)}"
        await ensure_item(
            slug,
            "badge",
            f"{tier_label} — {season_name}",
            description=f"Finished {tier_label.lower()} in {season_name}",
            asset=f"/badgeicons/{slug}.png",
            category="seasonal",
            season=season_name,
            rarity="legendary" if tier_label == "Champion" else "epic",
        )
        result = await grant_item(player_name, slug, "system:season-reset")
        if result == "granted":
            awarded.append(f"{player_name} — {tier_label}")
    return awarded
